precompiles/distribution/distribution.go

```diff
@@ -151,25 +157,69 @@ func (p Precompile) setWithdrawAddress(ctx sdk.Context, method *abi.Method, call
 }
 
 func (p Precompile) withdrawDelegationRewards(ctx sdk.Context, method *abi.Method, caller common.Address, args []interface{}, value *big.Int) ([]byte, error) {
-	if err := pcommon.ValidateNonPayable(value); err != nil {
+	err := p.validateInput(value, args, 1)
+	if err != nil {
 		return nil, err
 	}
 
-	if err := pcommon.ValidateArgsLength(args, 1); err != nil {
+	delegator, err := p.getDelegator(ctx, caller)
+	if err != nil {
+		return nil, err
+	}
+	_, err = p.withdraw(ctx, delegator, args[0].(string))
+	if err != nil {
+		return nil, err
+	}
+	return method.Outputs.Pack(true)
+}
+
+func (p Precompile) validateInput(value *big.Int, args []interface{}, expectedArgsLength int) error {
+	if err := pcommon.ValidateNonPayable(value); err != nil {
+		return err
+	}
+
+	if err := pcommon.ValidateArgsLength(args, expectedArgsLength); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p Precompile) withdraw(ctx sdk.Context, delegator sdk.AccAddress, validatorAddress string) (sdk.Coins, error) {
+	validator, err := sdk.ValAddressFromBech32(validatorAddress)
+	if err != nil {
 		return nil, err
 	}
+	return p.distrKeeper.WithdrawDelegationRewards(ctx, delegator, validator)
+}
+
+func (p Precompile) getDelegator(ctx sdk.Context, caller common.Address) (sdk.AccAddress, error) {
 	delegator, found := p.evmKeeper.GetSeiAddress(ctx, caller)
 	if !found {
 		return nil, fmt.Errorf("delegator %s is not associated", caller.Hex())
 	}
-	validator, err := sdk.ValAddressFromBech32(args[0].(string))
+
+	return delegator, nil
+}
+
+func (p Precompile) withdrawMultipleDelegationRewards(ctx sdk.Context, method *abi.Method, caller common.Address, args []interface{}, value *big.Int) ([]byte, error) {
```

Comment:
having a variable-length input would require calculating gas consumption dynamically, so we need to replay `Run()` with `RunAndCalculateGas`, similar to https://github.com/sei-protocol/sei-chain/blob/main/precompiles/wasmd/wasmd.go#L132 Specifically,
- gas limit should be converted and set in `ctx` https://github.com/sei-protocol/sei-chain/blob/main/precompiles/wasmd/wasmd.go#L145-L150
- remaining gas should be converted and returned from `ctx` https://github.com/sei-protocol/sei-chain/blob/main/precompiles/wasmd/wasmd.go#L258-L259

Right now we don't have a way to mix static and dynamic gas functions in the same precompile, so we change this new one to dynamic, we need to make the two existing functions dynamic as well

Reply:
Makes sense. Updated code to use dynamic gas calculation